[FEAT] 웹소캣 연결 및 송수신 기능 구현

kimyou1102 opened the pull request:
## Issue Number
closed #825

현재 pr에는 구현 방법이랑 구현의 이유에 대해서 적어뒀습니다! 더 자세한 내용은 노션 문서화에 정리해둘게요🙂

## As-Is
<!-- 문제 상황 정의 -->
- 웹소캣 연결이 안되어있는 상황

## To-Be
<!-- 변경 사항 -->
- 로그인 성공 시 응답값으로 memberId를 받아서 로컬스토리지에 저장
- stomp.js로 웹소캣 연결 및 송신 수신 기능 구현
- 메시지 보낼때 낙관적업데이트로 보낸 메시지가 바로 화면에 나타나게 구현

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?


## (Optional) Additional Description
### 웹소캣 연결 및 구독
Client 객체 생성
```
const client = new Client({
  webSocketFactory: () =>
    new SockJS(`${process.env.API_BASE_URL}/ws-chat`, null, {
      withCredentials: true,
    }),
  ...
});
```
- STOMP 클라이언트를 생성하여 웹소켓 연결을 준비
- SockJS를 통해 브라우저와 서버 간 호환성을 갖춘 웹소켓 연결 생성
- withCredentials: true → 쿠키나 인증 정보를 함께 전송

연결 성공 시 구독
```
onConnect: () => {
  client.subscribe(
    `/topic/chatroom/${chatRoomId}`,
    (message: IMessage) => {
      const parsedMessage = JSON.parse(message.body);

      setMessages((prev) =>
        prev.map((m) =>
          m.tempId && m.tempId === parsedMessage.tempId
            ? parsedMessage
            : m,
        ),
      );
    },
  );
},
```
- client.subscribe() 메서드를 통해 특정 토픽(/topic/chatroom/{chatRoomId}) 구독

### 메시지 송신
```
    client.publish({
      destination: `/app/chatroom/${chatRoomId}`,
      body: JSON.stringify({ content: message, chatRoomId, tempId }),
    });
```
- 사용자가 메시지를 입력하고 전송하면, /app/chatroom/{chatRoomId} 목적지로 메시지를 발행

### 낙관적 업데이트
메시지를 전송한 사용자에게 빠른 UI 처리를 제공하기 위해 낙관적 업데이트를 사용했어

1. 임시 메시지 생성 (`optimisticMsg`)
    - 사용자가 메시지 전송 버튼을 누르면, Date.now()를 사용하여 고유한 tempId (임시 ID)를 생성
    - 이 tempId와 함께 사용자 입력 내용(message), 보낸 사람 ID(memberId) 등을 포함한 임시 메시지 객체를 생성
    - 이 객체에는 전송 상태를 나타내는 status: 'pending' 속성을 추가
2. UI 즉시 업데이트
    - 생성된 임시 메시지 객체를 messages 상태 배열에 바로 추가하여 UI를 즉시 업데이트
    - 사용자는 자신의 메시지가 보내기 대기 중(pending) 상태로 화면에 표시되는 것을 즉시 확인 가능
3. 서버로 메시지 전송
    - UI 업데이트와 동시에, STOMP 클라이언트를 통해 실제 메시지 데이터를 서버로 전송(publish)
    - 이때, 앞서 생성한 tempId를 메시지 본문(body)에 포함하여 전송
4. 서버 응답 처리 및 상태 동기화
    - 서버는 메시지를 처리한 후, 수신한 tempId를 포함하여 구독 중인 토픽(/topic/chatroom/{chatRoomId})으로 메시지를 다시
    브로드캐스트
    - 클라이언트는 이 메시지를 수신하면, 메시지에 포함된 tempId를 사용해 messages 배열에서 해당 임시 메시지를 찾음
    - 찾은 임시 메시지를 서버로부터 받은 최종 데이터(실제 chatMessageId, createdAt 등 포함)로 교체. 이 과정을 통해 status도
    pending →  success 상태로 동기화

### 로그인 성공시 memberId를 로컬스토리지에 저장한 이유
- 수신한 메시지의 senderId와 비교해 내가 보낸 메시지인지 구분하고 말풍선 위치를 결정하기 위한 필수정보
- 채팅방 정보나 멘토링 정보 API에 포함시키는 것보다 로그인 시점에 저장하는 것이 자연스럽다고 판단
- memberId는 단순 식별자이므로 로컬스토리지에 저장해도 보안상 문제 없다고 판단되어서 이렇게 구현했어! 더 좋은 방법이 있으면 얘기해줘도 좋을 거 같아~~

## 추가된 기능 - 무한스크롤
무한스크롤에 관해 더 자세한 내용은 문서화에 정리해뒀습니다! 다음부터는 브랜치를 꼭 분리하겠습니다........

**봐야하는 파일**
- useInfiniteChatRoomMessage.ts
- useUpwardInfiniteScroll.ts
- ChatRoom.tsx
- chatRoom/mock 아래의 파일들


### 페이지네이션으로 변경 및 무한스크롤 적용
- useInfiniteQuery사용하여 페이지네이션 API 조회 - `useInfiniteChatRoomMessage`훅

### 상향식 무한스크롤 구현 
- 무한스크롤 중 새 채팅 추가 시 스크롤이 자동으로 맨 아래로 이동하던 문제를, 스크롤이 하단 근처에 있을 때만 이동하도록 수정
- 진입 시 스크롤이 자동으로 맨 아래로 이동하도록 변경하고, 해당 스크롤 동작이 완료된 후에만 옵저버가 등록되도록 수정 (useLayoutEffect로 구현)
  - 페이지 진입하자 마자 페이지네이션 동작하는 것을 방지
- IntersectionObserver 사용 - `useUpwardInfiniteScroll`훅
- 하향식과 다르게 상향식은 데이터가 추가되어도 컨테이너의 스크롤 위치가 그대로라 관찰 대상(타겟) 요소가 계속 가시 영역에 남아 IntersectionObserver가 연속으로 트리거되는 문제가 있음. 이를 위해 인식되지 않게 하기 위해 **데이터 추가 후 스크롤 조정**이 필요
  - fetchNextPage전에 이전 페이지 스크롤 상태 기록
  ```
        expectPrependRef.current = {
          prevH: list.scrollHeight,
          prevTop: list.scrollTop,
        };
  ```
  - 데이터 추가후 추가된 데이터만큼 스크롤 조정
  ```
  useLayoutEffect(() => {
    const list = listRef.current;
    const snap = expectPrependRef.current;
    if (!list || !snap) {
      return;
    }

    const delta = list.scrollHeight - snap.prevH;
    list.scrollTop = snap.prevTop + delta;

    expectPrependRef.current = null;
  }, [listRef, anchorKey]);
  ```

kimyou1102 committed:
feat: API_ENDPOINTS에 채팅 api url 추가 #825

kimyou1102 committed:
feat: getChatRooms 함수 생성 #825

kimyou1102 committed:
feat: getChatRooms API MSW 모킹 구현 #825

kimyou1102 committed:
feat: 이전 채팅 기록 불러오기 기능 구현 #825

kimyou1102 committed:
feat: 메시지 추가 시 자동으로 스크롤 맨 아래로 이동하는 기능 구현 #825

kimyou1102 committed:
refactor: Stompjs로 마이그레이션 #825

kimyou1102 committed:
refactor: Message 타입 변경 #825

kimyou1102 committed:
refactor: import문 순서 변경 #825

kimyou1102 committed:
refactor: 변경된 메시지 타입에 맞게 목데이터 변경 #825

kimyou1102 committed:
refactor: 로그인시 로컬스토리지에 memberId 저장 #825

kimyou1102 committed:
refactor: senderId를 로컬스토리지에 저장된 memberId를 사용하도록 변경 #825

kimyou1102 committed:
refactor: 웹소캣 주소 변경 #825

kimyou1102 committed:
fix: 채팅에 관한 API endpoint 오타 수정 #825

kimyou1102 committed:
refactor: memberId 상수화 되어있던 것을 로컬스토리지에서 받아오도록 변경 #825

kimyou1102 committed:
fix: map의 key값을 고유한 값으로 변경 #825

kimyou1102 committed:
refactor: 로컬스토리지에서 memberId만 가져오지 않고 전체 데이터 사용하도록 변경 #825

kimyou1102 committed:
feat: 전송 실패시 감지하는 로직 추가 #825

kimyou1102 committed:
refactor:: 메시지 전송 엔드포인트를 /app/chatroom로 변경 #825

kimyou1102 committed:
fix: WebSocket 연결 URL을 로컬에서 API_BASE_URL로 수정 #825

kimyou1102 committed:
refactor: Message.status 유니온 타입에 'pending' 추가 #825

kimyou1102 committed:
refactor: 낙관적 업데이트가 항상 화면에 표시되도록 WebSocket 연결 확인 조건문 순서 조정 #825

- 이전에는 연결이 안 되어 있으면 낙관적 업데이트조차 되지 않았음
- 조건문 순서를 바꿔 먼저 낙관적 업데이트 → 연결 확인 순으로 변경하여 UX 개선

kimyou1102 committed:
fix: 이전 채팅 기록 불러오는 API에 credentials 옵션 추가 #825

kimyou1102 committed:
feat: getChatRoomInfo 함수 생성 #825

kimyou1102 committed:
refactor: 명확한 함수명을 위해 getChatRooms를 getChatRoom으로 변경 #825

kimyou1102 committed:
feat: getChatRoomInfo API MSW 모킹 구현 #825

kimyou1102 committed:
refactor: 카멜케이스에 맞게 매개변수명 변경 #825

kimyou1102 committed:
refactor: getChatRoomInfo함수 반환 타입 변경 #825

kimyou1102 committed:
feat: 하드코딩한 채팅방 정보 api 응답값으로 변경 #825

kimyou1102 committed:
feat: 전송실패 UI 구현 #825

kimyou1102 committed:
refactor: form태그로 변경 및 onSubmit 핸들러 추가 #825

kimyou1102 committed:
refactor: 사용하지 않는 더미데이터 및 주석 제거 #825

kimyou1102 committed:
refactor: 센트리에 전송하는 에러 변경 #825

kimyou1102 committed:
refactor: 잘못된 전송 기능 제거  #825

kimyou1102 committed:
fix: tempId 기반 메시지 업데이트 시 status 'success' 적용 및 새 메시지 처리 개선 #825

kimyou1102 committed:
fix: 잘못 제거된 handleMessageSubmit 함수 다시 복구 #825

kimyou1102 committed:
build: sockjs-client, stompjs 설치 #825

kimyou1102 committed:
build: package-lock.json 삭제후 재설치 #825

dlsxjzld commented on frontend/public/mockServiceWorker.js:
p3: prettier 설정에 세미콜론이 true로 되어 있을 텐데 혹시 적용할 수 있을까? 🤔

kimyou1102 replied:
[refactor: 컨벤션에 맞게 세미콜론 추가](https://github.com/woowacourse-teams/2025-Fit-toring/pull/870/commits/ae0d9c7fb155b78fc5198a2cc340af59a82a4ccc)

dlsxjzld commented on frontend/src/common/mock/chatrooms/handlers.ts:
p4: msw 작성 좋다 👍

혹시 msw로 어떻게 채팅에 들어갈 수 있는지 방법도 써줄 수 있을까? 🤔

Beomtae replied:
p4: 나도 궁금해용

kimyou1102 replied:
msw 키고 /chat/room/{chatroomId} 로 들어가면 모킹한 이전 채팅 기록이 뜰거야!

dlsxjzld commented on frontend/src/pages/chatRoom/components/ChatBubble/ChatBubble.tsx:
```diff
@@ -1,27 +1,61 @@
+          <S_RetryIcon src={warningIcon} />
+          <S_RetryText>전송실패</S_RetryText>
+        </S_RetryInfoWrapper>
+      ) : null}
```
p4: 전송 실패일 때 뜨게 한 거 좋은데요~ 
그런데 스토리북을 보니까 전송 실패가 밑에 뜨고 있더라고! 
<img width="545" height="184" alt="스크린샷 2025-10-07 오후 9 27 29" src="https://github.com/user-attachments/assets/63915c71-0317-4d5a-a99b-fc595f2593e6" />

혹시 옆에 뜨게 하는 건 어떻게 생각해? 🧐 여러 개가 실패했을 때 메시지 사이의 간격이 많이 커질 것 같아서 말해봤어!

<img width="387" height="173" alt="스크린샷 2025-10-07 오후 9 22 41" src="https://github.com/user-attachments/assets/7b35b921-d31a-40c9-931f-23b7e2d6d46d" />

```tsx
function ChatBubble({ content, createdAt, authored, status }: ChatBubbleProps) {
  return (
    <S_Container authored={authored}>
      <S_BubbleWrapper authored={authored}>
        <S_Bubble authored={authored}>
          <S_Text authored={authored}>{content}</S_Text>
        </S_Bubble>
        <S_Temp>
          {status === 'fail' ? (
            <S_RetryInfoWrapper>
              <S_RetryIcon src={warningIcon} />
              <S_RetryText>전송실패</S_RetryText>
            </S_RetryInfoWrapper>
          ) : null}
          <S_Time>{formatToKoreanTime(createdAt)}</S_Time>
        </S_Temp>
      </S_BubbleWrapper>
    </S_Container>
  );
}

...

const S_Temp = styled.div`
  display: flex;
  flex-direction: column;
`;

const S_Time = styled.span`
  color: ${({ theme }) => theme.SYSTEM.GRAY400};
  text-align: end;
  ${({ theme }) => theme.TYPOGRAPHY.C5_R};
`;

```

kimyou1102 replied:
이런 디자인도 좋을 거 같아! 반영할게~~

kimyou1102 replied:
[style: 전송실패 메시지 위치 변경](https://github.com/woowacourse-teams/2025-Fit-toring/pull/870/commits/b00fa80f1993cee92a7abc44fcfb950b94fd8c83)

dlsxjzld commented on frontend/src/pages/login/components/LoginForm/LoginForm.tsx:
```diff
@@ -30,6 +30,15 @@ function LoginForm() {
   const fetchLogin = async () => {
     try {
       const response = await postLogin(userId, password);
+      const data = await response.json();
+
+      if (data?.memberId) {
+        localStorage.setItem(
+          'memberId',
+          JSON.stringify({ memberId: data.memberId }),
+        );
+      }
```
p3: data의 타입이 any로 잡혀서 as를 써서 잡아도 좋을 거 같은데 어떻게 생각해? 🤔 zod를 쓰는 방법도 있다는데 이건 좀 더 고민해보자!

kimyou1102 replied:
로건 PR에 보니까 post요청할때 제네릭 타입을 사용할 수 있게 해둬서 그렇게 수정해도 될거 같아!

zod를 사용하는게 제일 안전할 거 같긴한데 이건 외부라이브러리를 도입하는 거라 얘기를 해보고 도입해보면 좋을 거 같아!

현재는 이런식으로 수정하는 거 어떨까? 그러면 단언보다 더 안전하게 사용할 수 있을거 같아🙂

```
async post<T = void, R = void>({
    endpoint,
    body,
    withCredentials,
  }: ApiClientPostType<T>): Promise<R | undefined> {
    ...

    const sendRequest = async () => {
      const response = await fetch(url, options);
      if (!response.ok) {
        const data = await response.json();
        throw new ApiError(data.message, response.status);
      }

      try {
        return (await response.json()) as R;
      } catch {
        return undefined;
      }
    };

    return this.requestWithRefresh(sendRequest, endpoint);
  }
```

dlsxjzld replied:
try-catch로 감싸서 여기서 타입 단언을 시켜서 보내는 거 좋다!! 👍

그런데 이전에 사용처에서 response.status를 써야한다고 그대로 response를 return해야 한다고 했던 경우들이 있어서 await response.json()은 어려울 거 같아!

예를 들면 로그인 코드에서 response.status를 쓰고 있어서 어려울 거 같아!

```ts
if (data?.memberId) {
    localStorage.setItem(
      'memberId',
      JSON.stringify({ memberId: data.memberId }),
    );
  }

if (response.status === 200) {
  alert('로그인에 성공했습니다.');
  navigate(PAGE_URL.HOME);
  login();
}
```

이 문제를 해결하려면 post 메서드가 데이터와 response 객체를 둘 다 포함한 객체를 반환하는 방법도 있을 것 같아!! 
같이 고민해보면 좋을 거 같아~~

kimyou1102 replied:
아 맞아 그런 경우가 있어서 이부분은 더 얘기를 나눠보고 apiClient를 수정하는 게 좋을 거 같아! 
그리고 단언 보다는 어떤 값이 들어올지 몰라서 지금과 같이 타입가드를 사용하는게 더 안전할 거 같은데 어때?

dlsxjzld replied:
타입가드 좋은 거 같아!! 👍

dlsxjzld commented on frontend/src/pages/chatRoom/types/message.ts:
```diff
@@ -0,0 +1,14 @@
+  chatRoomId: number;
+  tempId: number;
+
+  // 서버에서 받는 id
+  chatMessageId?: number;
+
+  // 클라이언트의 상태 관리를 위한 속성
+  status?: 'success' | 'fail' | 'pending';
```
p5: 질문인데 얘네는 왜 옵셔널로 되어 있어? 🤔

kimyou1102 replied:
서버에서 주는 message 타입이랑 클라이언트에서 관리하는 message 타입이 달라서 주석으로 구분해뒀어.
- 서버에서 주는 message 타입: 서버·클라이언트 공통 필드 + 서버에서 받는 id
- 클라이언트에서 관리하는 message 타입: 서버·클라이언트 공통 필드 + 클라이언트 상태 관리를 위한 속성

status는 메시지를 보내는 중이거나 실패한 상태일 때 UI 처리를 하기 위해 필요해서 넣었고,
chatMessageId는 DB에 저장된 실제 고유 ID를 받아서 메시지를 매칭(map)할 때 쓰려고 추가했어.

서버에서 내려주는 구조랑 클라이언트에서 관리하는 구조가 거의 동일해서 타입을 분리하기보다 하나로 관리하는 게 더 깔끔하고 유지보수하기 좋다고 생각했어!

dlsxjzld replied:
하나로 합쳐서 관리하려는 의도가 뭔지 이해했어. 확실히 중복을 줄이는 좋은 방법인 것 같아! 👍 👍 

근데 듣고 보니, 꼭 필요한 값들인데도 옵셔널(?)로 되어 있어서 오히려 어떤 상황에서 어떤 값을 쓸 수 있는지 헷갈릴 수 있겠다 싶었어!

그래서 말인데, 아예 이렇게 공통 타입은 BaseMessage로 두고, ServerMessage랑 ClientMessage로 확장하는 건 어떨까?

```ts
// 1. 공통 속성을 Base 인터페이스로 추출
export interface BaseMessage {
  content: string;
  createdAt: string;
  senderId: number;
  chatRoomId: number;
  tempId: number;
}

// 2. 서버에서 받는 메시지 타입
export interface ServerMessage extends BaseMessage {
  chatMessageId: number; // 옵셔널이 아님
}

// 3. 클라이언트에서 관리하는 메시지 타입
export interface ClientMessage extends BaseMessage {
  status: 'success' | 'fail' | 'pending'; // 옵셔널이 아님
}
```

옵셔널이 사라지니까 ClientMessage에 status가 무조건 있다는 걸 보장받을 수 있고,
나중에 서버 타입만 바뀌어도 ServerMessage만 고치면 되니까 유지보수도 더 편해질 것 같아!

kimyou1102 replied:
messages 상태에는 클라이언트용 값이랑 서버용 값이 둘 다 공존해야 해서 하나의 타입을 쓸 수밖에 없어

내가 처음엔 분리할 수 있을 것처럼 설명했는데, 설명을 잘못한거 같아 미안해😥 
실제로는 완전히 분리할 수 없는 값이라 이대로 두는 게 맞는 것 같아

dlsxjzld commented on frontend/src/pages/chatRoom/ChatRoom.tsx:
```diff
@@ -133,25 +52,145 @@ function ChatRoom() {
 
+  const stompClientRef = useRef<Client | null>(null);
+
+  const { data } = useQuery({
+    queryKey: ['chatRoom', chatRoomId],
+    queryFn: () => getChatRoom(Number(chatRoomId!)),
+  });
+
+  const ChatRoomInfoQuery = useQuery<ChatRoomInfo>({
+    queryKey: ['chatRoomInfo', chatRoomId],
+    queryFn: () => getChatRoomInfo(Number(chatRoomId!)),
+  });
+
+  const chantRoomInfo = ChatRoomInfoQuery.data;
+  const mentoringId = chantRoomInfo?.mentoringId;
+
+  const { data: mentoring, isPending } = useQuery({
+    queryKey: ['mentoring', mentoringId],
+    queryFn: () => getMentoringDetail(String(mentoringId)),
+    enabled: !!mentoringId,
+  });
```
p2: useQuery 좋은데요~ 

코드를 보니까 useQuery 결과를 처리하는 방식이 두 가지가 함께 쓰이고 있는 것 같아.

하나는 { data }처럼 구조 분해하는 방식인데, 이때 data를 { data: chatRoomMessages }처럼 이게 무슨 데이터인지 명확히 알 수 있도록 네이밍 해주면, 나중에 코드를 볼 때 어떤 쿼리에서 온 데이터인지 바로 파악할 수 있어서 직관적인 것 같아. 다만 다른 useQuery의 결과가 있을 때 속성이 겹치면 네이밍을 해줘야 해서 조금 번거로울 수 있을 것 같아!

다른 하나는 ChatRoomInfoQuery처럼 객체 전체를 변수에 할당하는 방식이야! 이 방식은 ChatRoomInfoQuery.data나 ChatRoomInfoQuery.isLoading으로 쓰기 때문에 data나 isLoading 같은 속성 이름이 다른 쿼리랑 겹쳐도 신경 쓸 필요가 없고, isFetching이나 status 같은 다른 상태 값들을 쓰기 편할 것 같아!

두 방식 다 장단점이 명확해서 어느 쪽이든 한 가지로 통일하면 앞으로 코드를 읽거나 새로 추가할 때 훨씬 편해질 것 같아!

한 가지 방향으로 통일하면 훨씬 좋을 것 같은데 어떻게 생각해? 🧐

Beomtae replied:
p5: 개인적인 생각이지만 하나의 페이지에서 여러 API를 사용하는 경우도 있기 때문에 두번째 방식이 더 괜찮은 것 같아. 물론 ChatRoomInfoQuery.data 같이 좀 길어 질수는 있지만... ㅎㅎ 다들 어떻게 생각해???

dlsxjzld replied:
나도 범태 의견에 끌려~~

kimyou1102 replied:
난 첫번째 방식으로 하는거도 괜찮을거 같아 ChatRoomInfoQuery 범태 말대로 네이밍이 좀 길어질거라는 생각이 들어서 그런데 이건 취향차이라서 의견이 통일된다면 무슨 방법이든 상관없을 거 같아~~

dlsxjzld commented on frontend/src/pages/chatRoom/ChatRoom.tsx:
```diff
@@ -133,25 +52,145 @@ function ChatRoom() {
+    enabled: !!mentoringId,
+  });
+
+  useEffect(() => {
+    if (data) {
+      setMessages(data);
+    }
+  }, [data]);
```
p3: data가 useQuery에서 나온 값일텐데 무슨 데이터인지 네이밍이 되어 있으면 좋을 거 같아! 

setMessages를 보고 이 data가 messages와 관련된 데이터라는 걸 여기서 알 수 있어서 네이밍 처리가 있으면 좋을 것 같아 ㅎㅎ

kimyou1102 replied:
[refactor: 직관적인 이름을 위해 data 구조분해할당 시 별칭 지정(chantRoomMessage)](https://github.com/woowacourse-teams/2025-Fit-toring/pull/870/commits/40ab4ed3d501d47317b1e7977978265cb001581f)

dlsxjzld commented on frontend/src/pages/chatRoom/ChatRoom.tsx:
```diff
@@ -133,25 +52,145 @@ function ChatRoom() {
+
+            setMessages((prev) => {
+              if (parsedMessage.tempId) {
+                const index = prev.findIndex(
+                  (m) => Number(m.tempId) === Number(parsedMessage.tempId),
+                );
+                if (index !== -1) {
+                  const newArr = [...prev];
+                  newArr[index] = {
+                    ...parsedMessage,
+                    tempId: prev[index].tempId,
+                    status: 'success',
+                  };
+                  return newArr;
+                }
+              }
+
+              const exists = prev.some(
+                (m) =>
+                  m.chatMessageId &&
+                  m.chatMessageId === parsedMessage.chatMessageId,
+              );
+              if (exists) {
+                return prev;
+              }
+
+              return [...prev, { ...parsedMessage, status: 'success' }];
```
p4: 질문이 2개 있어!
1. 98번째 라인부터 111번째 라인까지는 낙관적 업데이트를 위한 임시 메시지 업데이트 맞을까? 
2. 113번째 라인부터 122번째 라인까지는 뭘 의미하는지 궁금해! 임시 메시지 업데이트와 무슨 차이가 있어??

Beomtae replied:
p4: 나도 113~122 라인의 코드가 무슨 역할을 하는지 잘 모르겠어 궁금해!!

kimyou1102 replied:
1. 맞아! 이미 낙관적 업데이트로 추가한 값을 찾아서 status만 바꾸는 역할이야
2. 이부분은 이미 담긴 값을 중복으로 담지 않기 위한 역할이야
chatMessageId는 서버에서 만든 고유한 값이기 때문에 이게 있다는 건 이미 서버에서 내려온 메시지가 있다는 뜻이라 바로 prev를 반환해줬어. 
chatMessageId가 없으면 parsedMessage와 일치하는 값을 서버에서 내려주지 않았다는 뜻이고, 서버에서 온 새로운 메시지이기 때문에 배열에 parsedMessage 추가한다는 뜻이야!

dlsxjzld replied:
> 이부분은 이미 담긴 값을 중복으로 담지 않기 위한 역할이야
chatMessageId는 서버에서 만든 고유한 값이기 때문에 이게 있다는 건 이미 서버에서 내려온 메시지가 있다는 뜻이라 바로 prev를 반환해줬어.
chatMessageId가 없으면 parsedMessage와 일치하는 값을 서버에서 내려주지 않았다는 뜻이고, 서버에서 온 새로운 메시지이기 때문에 배열에 parsedMessage 추가한다는 뜻이야!

여기서 궁금한 게, 혹시 서버에서 내려온 메시지가 있어야 낙관적 업데이트(1번)에서 status를 success로 업데이트 한 거 아니였어? 왜 한 번 더 서버에서 내려온 메시지가 있다고 확인하고 있어? 🤔

kimyou1102 replied:
첫 번째 중복 방지는 내가 낙관적 업데이트로 추가한 메시지만 대상으로 해.
tempId는 클라이언트에서 임시로 만든 메시지 객체에만 있고,
서버가 이걸 그대로 echo해서 다시 보내주면
그때 로컬에 있던 tempId가 같은 메시지를 찾아서 상태만 'success'로 바꿔줘.
내가 보낸 메시지가 두 번 쌓이지 않게 막는 역할이야.

두 번째 중복 방지는 서버가 같은 메시지를 여러 번 보내는 상황(예: 재연결, 재구독 등)을 대비한 거야.
이미 같은 chatMessageId를 가진 메시지가 있으면 새로 추가하지 않고,
서버 메시지가 중복으로 들어오는 걸 방지하는 역할로 뒀어!

dlsxjzld replied:
서버가 같은 메시지를 여러 번 보내는 상황을 대비한거구나!! 이해했어! 👍👍👍

정리하자면 

1번은 서버에서 보낸 메시지를 보고 낙관적 업데이트한 메시지를 업데이트 하기.
2번은 서버가 같은 메시지를 여러 번 보내는 상황을 대비하기.
구나!

맥락이 잘 담긴 함수로 빼도 좋겠다 👍👍
설명 고마워 😊

kimyou1102 replied:
내가 설명이 부족했던 거 같아🥲 노션에도 이내용 더 추가해둘게!! 함수로 빼는거도 좋을거 같아!

dlsxjzld commented on frontend/src/pages/chatRoom/ChatRoom.tsx:
```diff
@@ -133,25 +52,145 @@ function ChatRoom() {
+  useEffect(() => {
+    const client = new Client({
+      webSocketFactory: () =>
+        new SockJS(`${process.env.API_BASE_URL}/ws-chat`, null, {
+          withCredentials: true,
+        }),
```
p4: 혹시 SockJS.options에 withCredentials는 없는데 추가된 이유가 있을까? 이 값이 있어야 사용자를 인지할 수 있는 걸까?

혹시 이 값이 있어야 한다면 emotion styled에서 인터페이스 선언 병합한 것처럼 타입을 잡아주면 될 거 같아!

kimyou1102 replied:
서로 다른 도메인인 경우 저값을 true로 해줘야 쿠키가 전송된다고 알고있어! 로컬에서 테스트할때는 서로 다른 도메인이기 때문에 저옵션을 작성해줬어. 타입은 급하게 올린다고 해결을 못하고 올렸는데 반영할게!

kimyou1102 replied:
[feat: SockJS Options에 withCredentials 타입 추가](https://github.com/woowacourse-teams/2025-Fit-toring/pull/870/commits/9d5d67a0d54b23fd9e152d547cac0fdf120c503e)

dlsxjzld commented on frontend/src/pages/chatRoom/ChatRoom.tsx:
```diff
@@ -133,25 +52,145 @@ function ChatRoom() {
+                  const newArr = [...prev];
+                  newArr[index] = {
+                    ...parsedMessage,
+                    tempId: prev[index].tempId,
```
p4: 혹시 index !== -1이면 tempId가 일치한다는 건데, 이 코드가 있는 이유가 타입 때문에 일부러 남겨둔 걸까? 🤔

kimyou1102 replied:
클라이언트에서 넘겨준 tempId를 서버가 echo하는 거라 아래 부분을 삭제해도 될거 같아! 꼼꼼한 리뷰 고마워~~
```
tempId: prev[index].tempId,
```

kimyou1102 replied:
[refactor: tempId 기반 메시지 성공 처리 코드 단순화](https://github.com/woowacourse-teams/2025-Fit-toring/pull/870/commits/ae3f80cb763aea0ee0694801074dd5d1654668a5)

dlsxjzld reviewed (commented):
고생했어!!! 되게 깔끔하게 잘 만들었다 👍👍

낙관적 업데이트에 대한 흐름을 엄청 상세히 적어줘서 useQuery를 썼는데도 왜 useEffect와 상태가 필요한지 알 수 있었어!! 😊

msw를 잘 만들어놨는데 PR에 같이 적어두면 좋을 거 같아!! 👍

그리고 노션을 보니까 멘티가 멘토링 상세 페이지에서 채팅방으로 이동하기를 할 수 있는 것 같던데, 혹시 참여한 멘토링에서도 바로 이동할 수 있는걸까?

Beomtae reviewed (approved):
채팅방 구현하느라 고생많았어!! 로건이 피드백을 꼼꼼히 잘해줘서 더 남겨줄만한 코멘트는 없는 것 같아. 나도 궁금한 부분에 대한 질문만 남겨두었으니 확인 한번 부탁해 고생했어~

kimyou1102 committed:
refactor: 컨벤션에 맞게 세미콜론 추가 #825

kimyou1102 committed:
refactor: 직관적인 이름을 위해 data 구조분해할당 시 별칭 지정(chantRoomMessage) #825

kimyou1102 committed:
refactor: tempId 기반 메시지 성공 처리 코드 단순화 #825

- 서버에서 tempId를 echo해주는 것이기 때문에 명시적으로 지정해주지 않아도 되어서 제거

kimyou1102 committed:
refactor: 변경된 api 명세에 맞게 채팅 송신시 chatRoomId 바디에서 제거 #825

kimyou1102 committed:
style: 전송실패 메시지 위치 변경 #825

kimyou1102 committed:
feat: SockJS Options에 withCredentials 타입 추가 #825

kimyou1102 committed:
refactor: 변경된 api 명세에 맞게 모킹 데이터 및 타입 변경 #825

kimyou1102 committed:
refactor: 변경된 api명세에 맞게 api 한개만 사용하도록 변경 #825

- 채팅방 정보 조회 api에서 멘토링 정보까지 응답 받도록 변경되어서 멘토링 조회 api 사용 제거

kimyou1102 committed:
refactor: 페이지네이션 방식에 맞게 모킹데이터 및 msw 핸들러 함수 변경 #825

kimyou1102 committed:
refactor: 페이지네이션 방식에 맞게 getChatRoom API 함수 변경 #825

kimyou1102 committed:
refactor: ChatRoom 경로 파라미터 이름을 rooId에서 chatRoomId 로 수정 #825

kimyou1102 committed:
refactor: getChatRoom API험수 응답타입 변경 #825

kimyou1102 committed:
refactor: getChatRoom 호출을 useInfiniteQuery로 변경 #825

kimyou1102 committed:
refactor: 채팅방 진입 시 화면이 하단에서 시작되도록 변경 #825

- 렌더링 후 스크롤 이동 대신 처음부터 하단에서 렌더링되도록 수정하여
화면 깜빡임 없이 자연스럽게 최근 메시지를 확인할 수 있도록 개선

kimyou1102 committed:
refactor: 무한스크롤 적용 #825

kimyou1102 committed:
fix: 무한스크롤로 데이터 추가될때 스크롤 아래로 이동하는 문제 해결 #825

- 첫 렌더시에만 아래로 스크를 이동하도록 useLayoutEffect문 수정
- 추가되는 메시지를 위해 스크롤이 아래에 있을때 메시지가 추가된 경우만 스크롤이 아래로 이동하도록 useEffect문 추가

kimyou1102 committed:
fix: 위로 스크롤 시 IntersectionObserver 중복 트리거로 인한 연속 페이징 문제 해결 #825

- 스크롤이 최상단에 있어서 계속 트리거가 됨
- 데이터 추가시에 추가된 데이터만큼 스크롤 아래로 조정

kimyou1102 committed:
feat: 마지막 페이지 도달 시 추가 fetch 방지 로직 추가 #825

kimyou1102 committed:
refactor: 의존성 배열에 불필요한 요소들 제거 #825

kimyou1102 committed:
fix: fetchNextPage가 옵저버 위치 인식시에 여러번 호출되던 문제 해결 #825

kimyou1102 committed:
refactor: 채팅 메시지 페이지네이션 조회 로직을 useInfiniteChatRoomMessage 훅으로 분리 #825

kimyou1102 committed:
refactor: 채팅 메시지 페이지네이션 조회 로직 useInfiniteChatRoomMessage훅 적용 #825

kimyou1102 committed:
refactor: 무한 스크롤 조건문을 얼리 리턴 방식으로 변경 #825

kimyou1102 committed:
fix: useEffect랑 useLayoutEffect 바뀐 부분 수정 #825

kimyou1102 committed:
refactor: 이전 채팅 기록 무한스크롤 로직 훅으로 분리 #825

kimyou1102 committed:
refactor: 이전 채팅 기록 무한스크롤 useInfiniteChatRoomMessage훅 적용 #825

kimyou1102 committed:
refactor: useInfiniteChatRoomMessage훅 리턴값으로 useInfiniteQuery 자체를 리턴하게 변경 #825

kimyou1102 committed:
refactor: fetchNextPage 에러처리를 위해 try catch문 추가 #825

kimyou1102 committed:
refactor: 채팅 이전 기록 API 모킹 데이터 및 handler함수 수정 #825

kimyou1102 committed:
refactor: MessageResponse 타입 변경 #825

kimyou1102 committed:
refactor: 사용하지 않는 모킹데이터 제거 #825

kimyou1102 committed:
refactor: 채팅방 정보 조회API 반환값 변경에 따라 모킹 데이터 구조 변경 및 타입 변경 #825

kimyou1102 committed:
refactor: 채팅방 메시지 조회 API 쿼리 파라미터 변경 반영 #825

kimyou1102 committed:
refactor: 이전 채팅 조회 API 함수이름 더 명확하게 변경 getChatRoom -> getChatRoomMessages #825

kimyou1102 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2025-Fit-toring into feature/825-websocket-send-receive
